database/prefixdb/db.go

```diff
@@ -38,40 +45,69 @@ func NewNested(prefix []byte, db database.Database) *Database {
 	return &Database{
 		dbPrefix: hashing.ComputeHash256(prefix),
 		db:       db,
+		bufferPool: sync.Pool{
+			New: func() interface{} {
+				return make([]byte, 0, defaultBufCap)
+			},
+		},
 	}
 }
 
 // Has implements the Database interface
+// Assumes that it is OK for the argument to db.db.Has
+// to be modified after db.db.Has returns
 func (db *Database) Has(key []byte) (bool, error) {
 	db.lock.RLock()
 	defer db.lock.RUnlock()
 
 	if db.db == nil {
 		return false, database.ErrClosed
 	}
-	return db.db.Has(db.prefix(key))
+	prefixedKey := db.prefix(key)
+	defer func() {
+		prefixedKey = prefixedKey[:0]
+		db.bufferPool.Put(prefixedKey)
+	}()
+
+	return db.db.Has(prefixedKey)
 }
 
 // Get implements the Database interface
+// Assumes that it is OK for the argument to db.db.Get
+// to be modified after db.db.Get returns
 func (db *Database) Get(key []byte) ([]byte, error) {
 	db.lock.RLock()
 	defer db.lock.RUnlock()
 
 	if db.db == nil {
 		return nil, database.ErrClosed
 	}
-	return db.db.Get(db.prefix(key))
+	prefixedKey := db.prefix(key)
+	defer func() {
```

Comment:
Does deferring this function with these variables allocate more memory? Also I think I remember hearing that the performance of defer isn't great... Not sure that it matters enough to change it, but we might want to benchmark it.